packages/next/src/client/components/layout-router.tsx

```diff
@@ -389,6 +389,9 @@ function InnerLayoutRouter({
 
         return serverResponse
       })
+
+      // Suspend while waiting for lazyData to resolve
+      use(lazyData)
```

Comment:
this is the fix I added, though it's strange that this was necessary and feels like a potential React bug

Reply:
From offline discussion:

This makes sense since it just restores what we had in https://github.com/vercel/next.js/pull/66578/files#diff-2b58d532bd4381d1b195e96a465c9a90eef21cdd74e8ba12d8158f8eb2b54687L437 which also broke this behavior.

The issue is that by wrapping the page in a custom Suspense boundary, `InnerLayoutRouter` will be a child of that. However, `changeByServerResponse` will re-render the top level router and not necessarily the `InnerLayoutRouter` which hung indefinitely on `unresolvedThenable`.